File: python/1-D Dynamic Programming/partition_equal_subset_sum.py
from typing import List

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        # If the total sum is odd, we cannot partition into two equal subsets
        if sum(nums) % 2:
            return False
        
        # dp: set of all possible subset sums
        dp = set()
        dp.add(0)
        target = sum(nums) // 2

        # Iterate through the numbers backwards
        for i in range(len(nums) - 1, -1, -1):
            nextDP = set()
            for t in dp:
                # If we can reach the target sum, return True
                if (t + nums[i]) == target:
                    return True
                # Add both possibilities: including or excluding nums[i]
                nextDP.add(t + nums[i])
                nextDP.add(t)
            dp = nextDP
        # After processing all numbers, check if target is in dp
        return True if target in dp else False

# Time Complexity: O(n * s), where n is the number of elements and s is the target sum (sum(nums)//2).
# Space Complexity: O(s), for the set of possible subset sums.

# -------------------
# Test Cases
# -------------------
def test_partition_equal_subset_sum():
    s = Solution()
    # Test case 1: Example from Leetcode
    assert s.canPartition([1,5,11,5]) == True  # [1,5,5] and [11]
    # Test case 2: Cannot partition
    assert s.canPartition([1,2,3,5]) == False
    # Test case 3: All elements same
    assert s.canPartition([2,2,2,2]) == True
    # Test case 4: Single element
    assert s.canPartition([1]) == False
    # Test case 5: Empty list
    assert s.canPartition([]) == True  # Trivially partitionable
    # Test case 6: Large input, possible
    assert s.canPartition([100,100]) == True
    print("All test cases passed!")

if __name__ == "__main__":
    test_partition_equal_subset_sum()

File: C++/1-D Dynamic Programming/partition_equal_subset_sum.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        /*
            Problem:
            - Determine if the array can be partitioned into two subsets with equal sum.

            Approach:
            - If the total sum is odd, it’s impossible to split into two equal parts.
            - Use a set `dp` to store all possible subset sums as we iterate.
            - At each step, consider including or excluding the current number.
            - If at any point we can form `target = totalSum/2`, return true.

            Time Complexity: O(n * target)
                - At each of n steps, we generate up to target sums.
            Space Complexity: O(target)
                - The dp set stores sums up to target.
        */

        int total = accumulate(nums.begin(), nums.end(), 0);

        // If total sum is odd, cannot partition into equal halves
        if (total % 2 != 0) return false;

        int target = total / 2;

        unordered_set<int> dp;
        dp.insert(0);

        for (int i = nums.size() - 1; i >= 0; i--) {
            unordered_set<int> nextDP;

            for (int t : dp) {
                if (t + nums[i] == target) {
                    return true;
                }
                nextDP.insert(t);
                nextDP.insert(t + nums[i]);
            }

            dp = nextDP;
        }

        return dp.find(target) != dp.end();
    }
};

/*
================ LINE-BY-LINE SYNTAX EXPLANATION =================

#include <bits/stdc++.h>
    - Includes all standard C++ libraries (vector, set, algorithm, etc.)

using namespace std;
    - Allows you to write vector, string, etc. without the std:: prefix.

class Solution {
    - Defines a class named Solution.

public:
    - Marks the following members as publicly accessible.

bool canPartition(vector<int>& nums)
    - Function returns a boolean and takes a reference to a vector of integers.

int total = accumulate(nums.begin(), nums.end(), 0);
    - Calculates the sum of all elements in nums using the standard library.

if (total % 2 != 0) return false;
    - If the sum is odd, return false because it cannot be split evenly.

int target = total / 2;
    - The desired sum for one subset.

unordered_set<int> dp;
    - A set to keep track of all possible sums we can form.

dp.insert(0);
    - Start with sum 0 (empty subset).

for (int i = nums.size() - 1; i >= 0; i--)
    - Iterate backwards over the array.

unordered_set<int> nextDP;
    - Temporary set to store the next state of possible sums.

for (int t : dp)
    - For each sum already possible in dp.

if (t + nums[i] == target)
    - If adding current number to t equals target, return true.

nextDP.insert(t);
nextDP.insert(t + nums[i]);
    - Add both options: without or with current number.

dp = nextDP;
    - Update dp with the next state.

return dp.find(target) != dp.end();
    - After processing all elements, check if target is achievable.

*/
